File: clients/android/app/src/main/kotlin/com/lextures/android/core/lms/CourseImportExportLogic.kt
package com.lextures.android.core.lms

import com.lextures.android.core.network.ApiError
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.jsonObject

/** Course JSON backup/restore helpers (M13.10). */
object CourseImportExportLogic {
    enum class ImportMode { erase, mergeAdd, overwrite }

    /** Avoid loading very large backups fully in memory on device. */
    const val MAX_IMPORT_BYTES: Int = 50 * 1024 * 1024

    private val json = Json { ignoreUnknownKeys = true }

    fun webImportExportPath(courseCode: String): String =
        "/courses/$courseCode/settings/import-export"

    fun exportFileName(courseCode: String): String = "$courseCode-course-export.json"

    fun parseImportFileText(text: String): JsonObject {
        if (text.toByteArray(Charsets.UTF_8).size > MAX_IMPORT_BYTES) {
            throw ImportExportError.FileTooLarge
        }
        val element = runCatching { json.parseToJsonElement(text) }.getOrElse {
            throw ImportExportError.InvalidJson
        }
        val obj = runCatching { element.jsonObject }.getOrElse {
            throw ImportExportError.InvalidObject
        }
        if (obj.isEmpty()) throw ImportExportError.InvalidObject
        return obj
    }

    fun userFacingError(error: Throwable): String = when (error) {
        is ImportExportError -> error.message ?: "import-export-error"
        is ApiError.HttpStatus -> error.message ?: "import-export-error"
        else -> error.message ?: "import-export-error"
    }

    sealed class ImportExportError(message: String) : Exception(message) {
        data object InvalidJson : ImportExportError("invalid-json")
        data object InvalidObject : ImportExportError("invalid-object")
        data object FileTooLarge : ImportExportError("file-too-large")
    }
}

File: clients/ios/Lextures/Core/LMS/CourseImportExportLogic.swift
import Foundation

/// Course JSON backup/restore helpers (M13.10).
enum CourseImportExportLogic {
    enum ImportMode: String, CaseIterable, Identifiable {
        case erase
        case mergeAdd
        case overwrite

        var id: String { rawValue }
    }

    enum OperationState: Equatable {
        case idle
        case exporting
        case importing
        case success(String)
        case error(String)
    }

    /// Avoid loading very large backups fully in memory on device.
    static let maxImportBytes = 50 * 1024 * 1024

    static func webImportExportPath(courseCode: String) -> String {
        "/courses/\(courseCode)/settings/import-export"
    }

    static func exportFileName(courseCode: String) -> String {
        "\(courseCode)-course-export.json"
    }

    static func importModeTitleKey(_ mode: ImportMode) -> String {
        switch mode {
        case .erase: return "mobile.courseSettings.importExport.mode.erase.title"
        case .mergeAdd: return "mobile.courseSettings.importExport.mode.mergeAdd.title"
        case .overwrite: return "mobile.courseSettings.importExport.mode.overwrite.title"
        }
    }

    static func importModeDetailKey(_ mode: ImportMode) -> String {
        switch mode {
        case .erase: return "mobile.courseSettings.importExport.mode.erase.detail"
        case .mergeAdd: return "mobile.courseSettings.importExport.mode.mergeAdd.detail"
        case .overwrite: return "mobile.courseSettings.importExport.mode.overwrite.detail"
        }
    }

    static func importConfirmMessageKey(_ mode: ImportMode) -> String {
        switch mode {
        case .erase: return "mobile.courseSettings.importExport.confirmMessage.erase"
        case .mergeAdd: return "mobile.courseSettings.importExport.confirmMessage.mergeAdd"
        case .overwrite: return "mobile.courseSettings.importExport.confirmMessage.overwrite"
        }
    }

    static func parseImportFileData(_ data: Data) throws -> [String: JSONValue] {
        if data.count > maxImportBytes {
            throw ImportExportError.fileTooLarge
        }
        let decoded: [String: JSONValue]
        do {
            decoded = try JSONDecoder().decode([String: JSONValue].self, from: data)
        } catch {
            throw ImportExportError.invalidJSON
        }
        guard !decoded.isEmpty else {
            throw ImportExportError.invalidObject
        }
        return decoded
    }

    static func encodeExportForShare(_ bundle: [String: JSONValue]) throws -> Data {
        let encoder = JSONEncoder()
        encoder.outputFormatting = [.prettyPrinted, .sortedKeys]
        return try encoder.encode(bundle)
    }

    static func userFacingError(_ error: Error) -> String {
        if let importError = error as? ImportExportError {
            return importError.localizedDescription
        }
        if let apiError = error as? APIError {
            return apiError.errorDescription ?? L.text("mobile.courseSettings.importExport.genericError")
        }
        return error.localizedDescription
    }

    enum ImportExportError: LocalizedError, Equatable {
        case invalidJSON
        case invalidObject
        case fileTooLarge

        var errorDescription: String? {
            switch self {
            case .invalidJSON:
                return L.text("mobile.courseSettings.importExport.invalidJson")
            case .invalidObject:
                return L.text("mobile.courseSettings.importExport.invalidObject")
            case .fileTooLarge:
                return L.text("mobile.courseSettings.importExport.fileTooLarge")
            }
        }
    }
}
